File: src/dlc/tree.py
from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def levelOrder(root: Optional[TreeNode]) -> List[List[int]]:
    """102. Binary Tree Level Order Traversal"""
    if not root: 
        return []

    q = deque([root])

    res = []
    while q:
        n = len(q)
        level = []
        for _ in range(n):
            node = q.popleft()
            level.append(node.val)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

        res.append(level)
    
    return res


def invertTree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """226. Invert Binary Tree"""
    if not root:
        return root
    else:
        return TreeNode(root.val, invertTree(root.right), invertTree(root.left))

File: cpp/src/tree.cpp
#include <vector>
#include <deque>
#include "dlc_core/tree.h"

TreeNode* invertTree(TreeNode* root) {  // 226. Invert Binary Tree
    if (!root) {
        return root;
    } else {
        return (new TreeNode(root->val, invertTree(root->right), invertTree(root->left)));
    }
}

std::vector<std::vector<int>> levelOrder(TreeNode* root) {  // 102. Binary Tree Level Order Traversal

  if (root==nullptr) return {};

  std::vector<std::vector<int>> res;
  std::deque<TreeNode*> q{root};

  while (!q.empty()){
    std::vector<int> level = {};
    size_t n = q.size();
    for (size_t i=0 ; i<n ; i++) {
      TreeNode* node = q.front(); 
      level.push_back(node->val);
      if (node->left) q.push_back(node->left);
      if (node->right) q.push_back(node->right);
      q.pop_front();
    }
    res.push_back(level);
  }
  return res;
}